### ISSUE TITLE
go.mod auto update causing build binary fail

### ISSUE THREAD

**Issue by chinglinwen** [kubernetes-sigs/apiserver-builder-alpha#506]
## background
ProjectB(apiserver-builder-alpha) referenced a ProjectA( operator-sdk,kube-builder), using v0.18.2 version.

go.mod file auto updated follwing packages to v0.18.2(when build), so I can't change go.mod back to old version.
```
ProjectB's go.mod ( should be v0.17.2 )
	k8s.io/api v0.18.2
	k8s.io/apimachinery v0.18.2
	k8s.io/apiserver v0.18.2
	k8s.io/client-go v0.18.2
```
## problem

```
$ apiserver-boot build executables --generate=false
I0609 22:59:55.995348   16770 build_executables.go:168] CGO_ENABLED=0
I0609 22:59:55.995401   16770 build_executables.go:178] go build -o bin/apiserver cmd/apiserver/main.go
# qianxin.com/cloud-manager/pkg/client/clientset_generated/clientset/typed/core/v1
pkg/client/clientset_generated/clientset/typed/core/v1/openstack.go:62:5: not enough arguments in call to c.client.Get().Namespace(c.ns).Resource("openstacks").Name(name).VersionedParams(&options, scheme.ParameterCodec).Do
        have ()
        want (context.Context)
pkg/client/clientset_generated/clientset/typed/core/v1/openstack.go:79:5: not enough arguments in call to c.client.Get().Namespace(c.ns).Resource("openstacks").VersionedParams(&opts, scheme.ParameterCodec).Timeout(timeout).Do
        have ()
        want (context.Context)
pkg/client/clientset_generated/clientset/typed/core/v1/openstack.go:96:8: not enough arguments in call to c.client.Get().Namespace(c.ns).Resource("openstacks").VersionedParams(&opts, scheme.ParameterCodec).Timeout(timeout).Watch
        have ()
        want (context.Context)
pkg/client/clientset_generated/clientset/typed/core/v1/openstack.go:106:5: not enough arguments in call to c.client.Post().Namespace(c.ns).Resource("openstacks").Body(openstack).Do
        have ()
        want (context.Context)
pkg/client/clientset_generated/clientset/typed/core/v1/openstack.go:119:5: not enough arguments in call to c.client.Put().Namespace(c.ns).Resource("openstacks").Name(openstack.ObjectMeta.Name).Body(openstack).Do
        have ()
        want (context.Context)
pkg/client/clientset_generated/clientset/typed/core/v1/openstack.go:135:5: not enough arguments in call to c.client.Put().Namespace(c.ns).Resource("openstacks").Name(openstack.ObjectMeta.Name).SubResource("status").Body(openstack).Do
        have ()
        want (context.Context)
pkg/client/clientset_generated/clientset/typed/core/v1/openstack.go:147:5: not enough arguments in call to c.client.Delete().Namespace(c.ns).Resource("openstacks").Name(name).Body(options).Do
        have ()
        want (context.Context)
pkg/client/clientset_generated/clientset/typed/core/v1/openstack.go:163:5: not enough arguments in call to c.client.Delete().Namespace(c.ns).Resource("openstacks").VersionedParams(&listOptions, scheme.ParameterCodec).Timeout(timeout).Body(options).Do
        have ()
        want (context.Context)
pkg/client/clientset_generated/clientset/typed/core/v1/openstack.go:176:5: not enough arguments in call to c.client.Patch(pt).Namespace(c.ns).Resource("openstacks").SubResource(subresources...).Name(name).Body(data).Do
        have ()
        want (context.Context)
F0609 22:59:56.925971   16770 build_executables.go:183] exit status 2
```

```
$ apiserver-boot version
I0609 23:32:58.807585   32037 version.go:57] Version: version.Version{ApiserverBuilderVersion:"v1.17.0", KubernetesVendor:"kubernetes-1.17.3", GitCommit:"2662bf4ed7b794e7b39b0d4e7f634f92b1fad78a", BuildDate:"2020-04-03-20:30:07", GoOs:"linux", GoArch:"amd64"}
$ controller-gen --version
Version: v0.3.0
```

**Comment by chinglinwen** [comment]
I use the following replace in go.mod file, a new error occurs

```
replace	k8s.io/api =>        k8s.io/api v0.17.3
replace	k8s.io/apimachinery =>  k8s.io/apimachinery v0.17.3
replace	k8s.io/apiserver => k8s.io/apiserver v0.17.3
replace	k8s.io/client-go => k8s.io/client-go v0.17.3
```

```
$ apiserver-boot build executables --generate=false
I0610 10:26:12.592998   23974 build_executables.go:168] CGO_ENABLED=0
I0610 10:26:12.593071   23974 build_executables.go:178] go build -o bin/apiserver cmd/apiserver/main.go
# k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/internal
../../../pkg/mod/k8s.io/apiserver@v0.17.3/pkg/endpoints/handlers/fieldmanager/internal/gvkparser.go:55:31: cannot use *typeSchema (type "sigs.k8s.io/structured-merge-diff/v3/schema".Schema) as type "sigs.k8s.io/structured-merge-diff/schema".Schema in field value
F0610 10:26:28.532551   23974 build_executables.go:183] exit status 2
```

**Comment by chinglinwen** [comment]
using following replace fixed it

```
replace	k8s.io/api =>        k8s.io/api v0.17.3
replace	k8s.io/apimachinery =>  k8s.io/apimachinery v0.17.3
replace	k8s.io/apiserver => k8s.io/apiserver v0.17.3
replace	k8s.io/client-go => k8s.io/client-go v0.17.3
replace k8s.io/kube-openapi =>	k8s.io/kube-openapi v0.0.0-20191107075043-30be4d16710a
replace	k8s.io/gengo => k8s.io/gengo v0.0.0-20190822140433-26a664648505
replace	k8s.io/utils => k8s.io/utils v0.0.0-20191114184206-e782cd3c129f
replace	sigs.k8s.io/controller-runtime => sigs.k8s.io/controller-runtime v0.5.1
```

```diff
$ git diff go.mod 
diff --git a/go.mod b/go.mod
index 6d497cd..a3172ba 100644
--- a/go.mod
+++ b/go.mod
@@ -22,17 +22,18 @@ require (
        github.com/spf13/pflag v1.0.5
        github.com/tmc/grpc-websocket-proxy v0.0.0-20190109142713-0ad062ec5ee5 // indirect
        golang.org/x/xerrors v0.0.0-20191011141410-1b5146add898 // indirect
-       k8s.io/api v0.17.3
-       k8s.io/apimachinery v0.17.3
-       k8s.io/apiserver v0.17.3
-       k8s.io/client-go v0.17.3
-       k8s.io/gengo v0.0.0-20190822140433-26a664648505
+       k8s.io/api v0.18.2
+       k8s.io/apimachinery v0.18.2
+       k8s.io/apiserver v0.18.2
+       k8s.io/client-go v0.18.2
+       k8s.io/gengo v0.0.0-20200114144118-36b2048a9120
        k8s.io/klog v1.0.0
        k8s.io/kube-aggregator v0.17.3
-       k8s.io/kube-openapi v0.0.0-20191107075043-30be4d16710a
-       k8s.io/utils v0.0.0-20191114184206-e782cd3c129f
+       k8s.io/kube-openapi v0.0.0-20200121204235-bf4fb3bd569c
+       k8s.io/utils v0.0.0-20200324210504-a9aa75ae1b89
+       qianxin.com/openstack-operator v0.0.0-00010101000000-000000000000
        sigs.k8s.io/apiserver-builder-alpha v1.17.0
-       sigs.k8s.io/controller-runtime v0.5.1
+       sigs.k8s.io/controller-runtime v0.6.0
        sigs.k8s.io/controller-tools v0.1.12 // indirect
        sigs.k8s.io/kubebuilder v1.0.8
        sigs.k8s.io/testing_frameworks v0.1.1
@@ -47,3 +48,13 @@ replace github.com/markbates/inflect => github.com/markbates/inflect v1.0.4
 replace github.com/kubernetes-incubator/reference-docs => github.com/kubernetes-sigs/reference-docs v0.0.0-20170929004150-fcf65347b256
 
 replace sigs.k8s.io/structured-merge-diff => sigs.k8s.io/structured-merge-diff v1.0.1-0.20191108220359-b1b620dd3f06
+
+
+replace        k8s.io/api =>        k8s.io/api v0.17.3
+replace        k8s.io/apimachinery =>  k8s.io/apimachinery v0.17.3
+replace        k8s.io/apiserver => k8s.io/apiserver v0.17.3
+replace        k8s.io/client-go => k8s.io/client-go v0.17.3
+replace k8s.io/kube-openapi => k8s.io/kube-openapi v0.0.0-20191107075043-30be4d16710a
+replace        k8s.io/gengo => k8s.io/gengo v0.0.0-20190822140433-26a664648505
+replace        k8s.io/utils => k8s.io/utils v0.0.0-20191114184206-e782cd3c129f
+replace        sigs.k8s.io/controller-runtime => sigs.k8s.io/controller-runtime v0.5.1
\ No newline at end of file
```

**Comment by yue9944882** [comment]
@chinglinwen there was an incompatible change involved in the kubernetes upstream where a bunch of client interfaces were refactored since 1.18+. this will be resolved after bumping the mod dependencies to 1.18

**Comment by chinglinwen** [comment]
Since, the replace directive in go.mod can fixed the problem. I'm okay with the current pkg versions.
